### PR TITLE
First pass at a permanently failed output stream from the spout.

### PR DESCRIPTION
**Pull request by Crim**
**Use Case**
We'd like the ability to handle tuples that are considered "permanently failed".  We're defining "permanently failed" meaning that the topology attempted to process the tuple at least once and the `RetryManager` implementation has determined that the tuple should never be retried.  When this occurs, the tuple will be emitted un-anchored out a "failed" stream.  Bolts within the topology can subscribe to this "failed" stream and do its own error handling.

It should be noted (again) that tuples emitted on the "failed" stream are un-anchored and should be considered unreliable.  IE. The spout will not track it or be notified of failed or acked tuples emitted down this "failed" stream.

**Notes**

One idea was to move the RetryFailedMessageManager out of the VirtualSpouts and only have a single instance living within DynamicSpout (instead of one per VSpout).  Doing so would simplify the logic around being able to easily identify and emit out permanently failed tuples.  The trade off is this alters how we enforce fairness across VirtualSpouts.  If we continue to give priority to failed tuples being replayed over new tuples means that a single VSpout emitting tons of tuples that end up failing could overwhelm the spout's output.

Instead I added an additional flag on the Message object that allows you to mark it as permanently failed, and DynamicSpout uses that flag to determine which stream to emit down.  Not entirely the most elegant solution, so open to other ideas.

Thoughts?

**Todo**
Update changelog & readme

### CODE REVIEW

File: src/main/java/com/salesforce/storm/spout/dynamic/DynamicSpout.java [hunk 1]
```diff
@@ -491,6 +511,16 @@ String getOutputStreamId() {
         return outputStreamId;
     }
 
+    /**
+     * @return The stream that tuples that have permanently failed will be emitted out.
+     */
+    String getPermanentlyFailedOutputStreamId() {
```
Comment (Crim) [inline src/main/java/com/salesforce/storm/spout/dynamic/DynamicSpout.java]: Assumption here is that we're just taking the standard output stream and tacking on "_failed".

Other option is to allow explicitly defining this value.

Or we could just hard code it to something like "permanently_failed"

Or some other option?

File: src/test/java/com/salesforce/storm/spout/dynamic/VirtualSpoutTest.java [hunk 2]
```diff
@@ -1656,9 +1764,7 @@ public void testSetEndingState() {
         defaultConfig.put(KafkaConsumerConfig.DESERIALIZER_CLASS, Utf8StringDeserializer.class.getName());
 
         // DynamicSpout config items
-        defaultConfig.put(SpoutConfig.PERSISTENCE_ZK_ROOT, "/sideline-spout-test");
```
Comment (Crim) [inline src/test/java/com/salesforce/storm/spout/dynamic/VirtualSpoutTest.java]: Removing the ZK dependency (not needed in this test) plus calling close makes this test run in less than 1 sec, compared to 1 min 30 sec+ previously. 🎈

File: src/main/java/com/salesforce/storm/spout/dynamic/DynamicSpout.java (hunk 1 above)
Comment (stanlemon) [reply src/main/java/com/salesforce/storm/spout/dynamic/DynamicSpout.java]: This seems like it should probably be driven by a configuration value with a decent default.

Comment (Crim) [reply src/main/java/com/salesforce/storm/spout/dynamic/DynamicSpout.java]: added SpoutConfig.PERMANENTLY_FAILED_OUTPUT_STREAM_ID => "spout.permanently_failed_output_stream_id" with a default value of "failed"

**Pull request merged**